operators/src/engine/operator.rs

```diff
@@ -65,21 +43,236 @@ pub trait VectorOperator: Operator + CloneableVectorOperator {
+            .into_iter()
+            .map(|o| o.initialized_operator(context))
+            .collect::<Result<Vec<Box<dyn InitializedRasterOperator>>>>()?;
+        let vector_sources = vector_sources_not_init
+            .into_iter()
+            .map(|o| o.into_initialized_operator(context))
+            .collect::<Result<Vec<Box<dyn InitializedVectorOperator>>>>()?;
+        let result_descriptor = result_descriptor_fn(
+            &params,
+            &context,
+            raster_sources.as_slice(),
+            vector_sources.as_slice(),
+        )?;
+
+        Ok(Self::new(
+            params,
+            context,
+            result_descriptor,
+            raster_sources,
+            vector_sources,
+        ))
+    }
+}
+
+impl<P, R> InitializedOperator for InitilaizedOperatorImpl<P, R>
+where
+    P: std::fmt::Debug + Clone,
+    R: std::fmt::Debug + Clone,
+{
+    fn execution_context(&self) -> &ExecutionContext {
+        &self.context
+    }
+    fn raster_sources(&self) -> &[Box<dyn InitializedRasterOperator>] {
+        self.raster_sources.as_slice()
+    }
+    fn vector_sources(&self) -> &[Box<dyn InitializedVectorOperator>] {
+        self.vector_sources.as_slice()
+    }
+    fn raster_sources_mut(&mut self) -> &mut [Box<dyn InitializedRasterOperator>] {
+        self.raster_sources.as_mut_slice()
+    }
+    fn vector_sources_mut(&mut self) -> &mut [Box<dyn InitializedVectorOperator>] {
+        self.vector_sources.as_mut_slice()
+    }
+}
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
+pub struct RasterResultDescriptor {
+    pub data_type: RasterDataType,
+    pub projection: ProjectionOption,
+}
+
+impl RasterResultDescriptor {
+    pub fn map<F>(self, f: F) -> Self
+    where
+        F: Fn(Self) -> Self,
+    {
+        f(self)
+    }
+
+    pub fn map_data_type<F>(mut self, f: F) -> Self
+    where
+        F: Fn(RasterDataType) -> RasterDataType,
+    {
+        self.data_type = f(self.data_type);
+        self
+    }
+
+    pub fn map_projection<F>(mut self, f: F) -> Self
+    where
+        F: Fn(ProjectionOption) -> ProjectionOption,
+    {
+        self.projection = f(self.projection);
+        self
+    }
+}
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
+pub struct VectorResultDescriptor {
+    pub data_type: VectorDataType,
+    pub projection: ProjectionOption,
+}
+
+impl VectorResultDescriptor {
+    pub fn map<F>(self, f: F) -> Self
+    where
+        F: Fn(Self) -> Self,
+    {
+        f(self)
+    }
+
+    pub fn map_data_type<F>(mut self, f: F) -> Self
+    where
+        F: Fn(VectorDataType) -> VectorDataType,
+    {
+        self.data_type = f(self.data_type);
+        self
+    }
+
+    pub fn map_projection<F>(mut self, f: F) -> Self
+    where
+        F: Fn(ProjectionOption) -> ProjectionOption,
+    {
+        self.projection = f(self.projection);
+        self
+    }
+}
```

Comment:
Maybe we should put this logic into a separate file

Reply:
ok